Added the document for LeaderElection in multi-node etcd (#407)
gardener/etcd-backup-restore: doc/concepts/leader_election.md

### Leading ETCD main container’s sidecar is the backup leader 

- The `backup-restore sidecar` poll its own etcd main container to see if it is the leading member in the etcd cluster.

> This information is used by the backup-restore sidecars to decide that sidecar of the leading etcd main container is the `backup leader`.

Reviewer: Consider replacing with the following
`The backup-restore sidecar corresponding to the leading etcd member becomes the backup leader while the other backup-restore sidecars become followers`

Author: here I'm just wanted to explain the about `backup leader` only.

Reviewer: Then maybe consider only the first part of the sentence

- Only `backup leader` sidecar among the members have the responsibility to take/upload the snapshots(full as well as incremental) for a given Etcd cluster as well as to [trigger the defragmentation](https://github.com/gardener/etcd-druid/tree/master/docs/proposals/multi-node#defragmentation) for each Etcd cluster member. 

### Work flow

Backup-restore can be in following 3 states:
1. Follower:

>    - DefaultState for every backup-restore member is `Follower` State.

Reviewer: ```suggestion
   -  The default state for every backup-restore member is the `Follower` State.
```

>    - If it's corresponding etcd main container is a `Follower` etcd.

Reviewer 2: Replace `if it's` with `when the`

Reviewer 2: Also, move the second point up and bring the first point of `default state` down.

2. Leader:

>    - If its own etcd main containers becomes the `leader` then sidecar will also become `leading sidecar`.

Reviewer: ```suggestion
   - If its corresponding etcd main containers becomes the `leader` then sidecar will also become `leading sidecar`.
```

3. Unknown:

>    - If there is no etcd leader present or Quorum loss.

Reviewer: ```suggestion
   - If there is no etcd leader present in the cluster or in the case of a quorum loss.
```

   - If corresponding etcd main container is down and [endpoint status](https://github.com/etcd-io/etcd/blob/f82b5cb7768dacad9fb310232c1383b4e6718378/client/v3/maintenance.go#L53) api call fails.

### Backup

- Only `backup leader` among the backup-restore members have the responsibility to take/upload the snapshots(full as well as incremental) for a given Etcd cluster. 

> - `backup leader` also have the responsibility to [garbage-collect](https://github.com/ishan16696/etcd-backup-restore/blob/leaderElection-doc/doc/usage/getting_started.md#taking-scheduled-snapshot) the backups according to configured garbage collection policy.

Reviewer: ```suggestion
- The `backup leader` also has the responsibility to [garbage-collect](https://github.com/ishan16696/etcd-backup-restore/blob/leaderElection-doc/doc/usage/getting_started.md#taking-scheduled-snapshot) backups from the cloud bucket according to a configured garbage collection policy.
```

Reviewer 2: Why garbage-collect is pointing to `taking scheduled snapshot` ?

Reviewer: The `taking scheduled snapshot` section mentions the garbage collection policies with small examples

Author: yes, @aaronfern  is correct there is no separate section for garbage collection policies, it is explain inside `taking scheduled snapshot`

### Member-lease

> Each backup-restore member have the responsibility to renew its member lease periodically, this will work as heartbeat and indicates that backup-restore cluster members are in `Healthy` State.

Reviewer: ```suggestion
Each backup-restore member has the responsibility to renew its member lease periodically. This is intended to simulate a heartbeat and indicate that the backup-restore cluster member is in a `Healthy` State.
```

### Defragmentation

> Defragmentation is performed only by the `leading backup-restore` sidecar. The defragmentation is performed only when etcd cluster is in full health and it is done in a rolling manner for each members to avoid disruption. Defragmentation is triggered on all etcd follower then at last on leading etcd member.

Reviewer: ```suggestion
Defragmentation for all cluster members is coordinated by the `leading backup-restore` sidecar. Defragmentation is performed only when etcd cluster is in full health and it is done in a rolling manner for each member to avoid disruption. Defragmentation is triggered first on all etcd follower members followed lastly by leading etcd member.
```

Reviewer 2: At first, `leading backup-restore` sidecar triggers defragmentation on all etcd follower members one by one and at last, on itself.
